### PR TITLE
Fix the issue of case sensitivity when matching keys with cache enabled

### LINKED ISSUE #3529: Apollo-ConfigService开启缓存后，过期缓存中配置需要校验大小写，会导致apollo-client填写APP_ID参数没有严格大小写的服务获取不到配置
- [x] 我已经检查过[discussions](https://github.com/ctripcorp/apollo/discussions)
- [x] 我已经搜索过[issues](https://github.com/ctripcorp/apollo/issues)
- [x] 我已经仔细检查过[FAQ](https://github.com/ctripcorp/apollo/wiki/%E9%83%A8%E7%BD%B2&%E5%BC%80%E5%8F%91%E9%81%87%E5%88%B0%E7%9A%84%E5%B8%B8%E8%A7%81%E9%97%AE%E9%A2%98)

**描述bug**
Apollo-ConfigService开启缓存功能后，客户端填写的APP_ID参数没有严格区分大小写的服务会获取不到最新的配置。
不启用服务端缓存，不会校验APP_ID大小写的。
**复现**

通过如下步骤可以复现：

1. apolloconfigdb.serverconfig表中字段config-service.cache.enabled设置成true，并重启Apollo-ConfigService服务
2. Apollo上有一个项目Apollo-Demo
3. 客户端有一个服务APP_ID=apollo-demo
4. 启动客户端后，再发布一个配置，会发现客户端一直获取不到最新的配置

有问题的代码片段（从Apollo-ConfigService模块ConfigServiceWithCache类中抽出）：
```
    //cache is out-dated
    if (clientMessages != null && clientMessages.has(key) &&
        clientMessages.get(key) > cacheEntry.getNotificationId()) {
      //invalidate the cache and try to load from db again
      invalidate(key);
      cacheEntry = configCache.getUnchecked(key);
    }
```
上述判断中has(key)实际是一个map的containsKey方法，大小写不同会导致比较失败，无法进入过期缓存的逻辑
**期望**
补充一个私有方法：
```
  boolean cacheOutOfDate(String key, ConfigCacheEntry cacheEntry, ApolloNotificationMessages clientMessages) {
    if (clientMessages == null || clientMessages.isEmpty()) {
      return false;
    }
    Iterator<Map.Entry<String, Long>> iterator = clientMessages.getDetails().entrySet().iterator();
    while (iterator.hasNext()) {
      Map.Entry<String, Long> detail = iterator.next();
      if (detail.getKey().toLowerCase().equals(key.toLowerCase()) && detail.getValue() > cacheEntry
              .getNotificationId()) {
        return true;
      }
    }
    return false;
  }
```
上面有问题的判断替换成私有的方法

```
    //cache is out-dated
    if (cacheOutOfDate(key, cacheEntry, clientMessages)) {
      //invalidate the cache and try to load from db again
      invalidate(key);
      cacheEntry = configCache.getUnchecked(key);
    }
```
**截图**

如果可以，附上截图来描述你的问题

### 额外的细节和日志

- 版本：
- 错误日志
- 配置：
- 平台和操作系统

### PR DESCRIPTION
**Pull request by klboke**
## What's the purpose of this PR

Fix the issue of case sensitivity when matching keys with cache enabled

- Add a configuration` config-service.cache.key.ignore-case` to control whether the cache key is case-sensitive.

## Which issue(s) this PR fixes:
Fixes #3529

### CODE REVIEW

File: apollo-configservice/src/test/java/com/ctrip/framework/apollo/configservice/service/config/ConfigServiceWithCacheAndCacheKeyIgnoreCaseTest.java [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+  }
+
+  @Test
+  public void testFindLatestActiveRelease() {
+    when(releaseMessageService.findLatestReleaseMessageForMessages(Lists.newArrayList(lowerCaseSomeKey)))
+        .thenReturn(someReleaseMessage);
+    when(releaseService.findLatestActiveRelease(
+        matches(Pattern.compile(someAppId, Pattern.CASE_INSENSITIVE)),
+        matches(Pattern.compile(someClusterName, Pattern.CASE_INSENSITIVE)),
+        matches(Pattern.compile(someNamespaceName, Pattern.CASE_INSENSITIVE))
```
Comment (mghio) [inline apollo-configservice/src/test/java/com/ctrip/framework/apollo/configservice/service/config/ConfigServiceWithCacheAndCacheKeyIgnoreCaseTest.java:165]: How about replacing it with this method?

```Java
private String matchesCaseInsensitive(final String regex) {
    return matches(Pattern.compile(regex, Pattern.CASE_INSENSITIVE));
}
```

**Review by mghio — CHANGES_REQUESTED** [review]

File: apollo-configservice/src/test/java/com/ctrip/framework/apollo/configservice/service/config/ConfigServiceWithCacheAndCacheKeyIgnoreCaseTest.java (hunk 1 above)
Comment (klboke) [reply apollo-configservice/src/test/java/com/ctrip/framework/apollo/configservice/service/config/ConfigServiceWithCacheAndCacheKeyIgnoreCaseTest.java:165]: good

**Comment by nobodyiam** [comment]
This attempt is commendable, but a problem may arise if the database is set to be case-sensitive. In such a scenario, the cache loader would fail to fetch data from the database, since the appid, cluster, and namespace have all been converted to lowercase.

**Comment by klboke** [comment]
> but a problem may arise if the database is set to be case-sensitive. In such a scenario, the cache loader would fail to fetch data from the database, since the appid, cluster, and namespace have all been converted to lowercase.

There is indeed such a situation, and this PR also considers it. It is currently not enabled by default, and needs to be enabled through a configuration switch. This has also been prominently explained in the documentation.

**Review by nobodyiam — APPROVED** [review]
LGTM

**Pull request merged**